### PR TITLE
[ADD] odoo_sh: How to debug on Odoo SH

### PR DESCRIPTION
**Pull request by jev-odoo**

### CODE REVIEW

File: odoo_sh/advanced/containers.rst [hunk 1]
```diff
@@ -186,3 +186,37 @@ Look for "*odoo: addons paths*":
 **Be careful**, especially with your production database.
 Operations that you perform running this Odoo server instance are not isolated:
 Changes will be effective in the database. Always, make your tests in your staging databases.
+
+Debugging in Odoo SH
+====================
+
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: I would add an introductory paragraph like this:

> Debbugging an Odoo.sh build is not really different than another Python app. This article only explains the specificities and limitations of the Odoo.sh platform, and assumes that you already know how to use a debugger. 
> If you don't know how to debug a Python application yet, there are multiple introductory courses that can be easily found on the Internet.

File: odoo_sh/advanced/containers.rst [hunk 2]
```diff
@@ -186,3 +186,37 @@ Look for "*odoo: addons paths*":
...
+You can use pdb or pudb to debug your code on Odoo SH.
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: Nitpicking on formatting: use backticks for `code/terminal output` things

Comment (rim-odoo) [reply odoo_sh/advanced/containers.rst]: ipdb too?

File: odoo_sh/advanced/containers.rst [hunk 3]
```diff
@@ -186,3 +186,37 @@ Look for "*odoo: addons paths*":
...
+As the server is run outside a shell, you cannot do it directly from your Odoo instance backend as the debugger needs a shell to operate.
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: > you cannot launch the debugger directly

File: odoo_sh/advanced/containers.rst [hunk 4]
```diff
@@ -186,3 +186,37 @@ Look for "*odoo: addons paths*":
...
+
+pdb is installed by default on all containers. If you want to use `pudb <https://pypi.org/project/pudb/>`_, you have to install it before:
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: ipdb ? 
add a link to pdb documentation?

File: odoo_sh/advanced/containers.rst [hunk 5]
```diff
@@ -186,3 +186,37 @@ Look for "*odoo: addons paths*":
...
+
+.. code-block:: bash
+
+  $  pip install pudb --user
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: or simpler: add pudb to the requirements.txt (even if it does not look professional in a production project ^^)

File: odoo_sh/advanced/containers.rst [hunk 6]
```diff
@@ -186,3 +186,37 @@ Look for "*odoo: addons paths*":
...
+
+Then edit the code where you want to insert the debugger and add this:
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: > you want to trigger? start? the debugger

File: odoo_sh/advanced/containers.rst [hunk 7]
```diff
@@ -186,3 +186,37 @@ Look for "*odoo: addons paths*":
...
+
+.. code-block:: python
+
+  import sys
+  if sys.__stdin__.isatty():
+    import pdb; pdb.set_trace()
+
+The condition is there so as it won't crash if you run the action from your Odoo instance backend.
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: > The condition `sys.__stdin__.isatty()` is a hack that detects if you run Odoo from a shell.

File: odoo_sh/advanced/containers.rst [hunk 8]
```diff
@@ -186,3 +186,37 @@ Look for "*odoo: addons paths*":
...
+
+Save the file and then run the Odoo Shell:
+
+.. code-block:: bash
+
+  $  odoo-bin shell
+
+Now, via the Odoo Shell, you can trigger the action to debug.
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: > you can trigger the piece of code/function/method you want to debug.

File: odoo_sh/advanced/containers.rst [hunk 9]
```diff
@@ -186,3 +186,37 @@ Look for "*odoo: addons paths*":
...
+
+.. image:: ./media/pdb_sh.png
+    :align: center
+    :alt: Text that is displayed when your image is not available (eg: for screen readers and other accessibility tools)
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: Well, add a description for your image ;)

**Comment by jev-odoo** [comment]
@rim-odoo Done

File: odoo_sh/advanced/containers.rst [hunk 10]
```diff
@@ -186,3 +186,53 @@ Look for "*odoo: addons paths*":
...
+Debugging in Odoo SH
+====================
+
+Debbugging an Odoo.sh build is not really different than another Python app. This article only explains the specificities and limitations of the Odoo.sh platform, and assumes that you already know how to use a debugger.
+If you don't know how to debug a Python application yet, there are multiple introductory courses that can be easily found on the Internet.
+
+You can use `pdb`, `pudb` or `ipdb` to debug your code on Odoo SH.
+As the server is run outside a shell, you cannot launch the debugger directly from your Odoo instance backend as the debugger needs a shell to operate.
+
+`pdb <https://docs.python.org/2/library/pdb.html>`_ is installed by default on all containers. If you want to use `pudb <https://pypi.org/project/pudb/>`_ or `ipdb <https://pypi.org/project/ipdb/>`_ you have to install it before.
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: Python 2 is outdated, you can add the link to P3 safely :nerd_face:

File: odoo_sh/advanced/containers.rst [hunk 11]
```diff
@@ -186,3 +186,53 @@ Look for "*odoo: addons paths*":
...
+
+You have two options:
+
+- temporary (will be available only in the build it has been installed in:
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: (Count (your (parentheses :laughing:  )))

File: odoo_sh/advanced/containers.rst [hunk 12]
```diff
@@ -186,3 +186,53 @@ Look for "*odoo: addons paths*":
...
+      $  pip install pudb --user
+
+    or
+
+    .. code-block:: bash
+
+      $  pip install ipdb --user
+
+- Permanent:
+    add `pudb` or `ipdb` in your project requirement.txt
```
Comment (rim-odoo) [inline odoo_sh/advanced/containers.rst]: `requirementS.txt`

**Comment by jev-odoo** [comment]
@rim-odoo Done again ;)

**Pull request merged**